### PR TITLE
Fixed issue #17301: tokens > minute lock out

### PR DESCRIPTION
**Pull request by gabrieljenik**
Added a whitelist as global security settings as to be able to whitelist certain ips either for token or login attempts.

### CODE REVIEW

File: application/models/FailedLoginAttempt.php [hunk 1]
```diff
@@ -126,4 +137,35 @@ public function addAttempt()
...
+     *
+     * @param string $ip
+     * @param string $attemptType   'login' or 'token'
+     *
+     * @return boolean
+     */
+    private function isWhitelisted($ip, $attemptType)
+    {
+        if ($attemptType != self::TYPE_LOGIN && $attemptType != self::TYPE_TOKEN) {
+            return false;
```
Comment (olleharstedt) [inline application/models/FailedLoginAttempt.php:152]: Either `ThrowInvalidArgument` or `assert`

**Review by olleharstedt — CHANGES_REQUESTED** [review]

**Comment by Shnoulle** [comment]
Not related to this commit : time for ipv6 ?

/me must report the feature request …

File: application/models/FailedLoginAttempt.php (hunk 1 above)
Comment (gabrieljenik) [reply application/models/FailedLoginAttempt.php]: Done

Comment (olleharstedt) [reply application/models/FailedLoginAttempt.php]: Hooray!

File: application/helpers/sanitize_helper.php [hunk 2]
```diff
@@ -502,3 +502,22 @@ function sanitize_signedint($integer, $min = '', $max = '')
...
+{
+    if ($allowWildcards) {
+        $regex = '/^((([0-9]{1,2})|(1[0-9]{2,2})|(2[0-4][0-9])|(25[0-5])|\*)\.){3}(([0-9]{1,2})|(1[0-9]{2,2})|(2[0-4][0-9])|(25[0-5])|\*)$/';
+    } else {
+        $regex = '/^((([0-9]{1,2})|(1[0-9]{2,2})|(2[0-4][0-9])|(25[0-5]))\.){3}(([0-9]{1,2})|(1[0-9]{2,2})|(2[0-4][0-9])|(25[0-5]))$/';
+    }
+    return preg_match($regex, $string, $m) ? $string : '';
+}
+
+function check_ip_address($input, $allowWildcards = true)
```
Comment (olleharstedt) [inline application/helpers/sanitize_helper.php:517]: Missing docblock :( And above

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: application/models/FailedLoginAttempt.php [hunk 3]
```diff
@@ -126,4 +137,34 @@ public function addAttempt()
...
+    /**
+     * Returns true if the specified IP is whitelisted
+     *
+     * @param string $ip
+     * @param string $attemptType   'login' or 'token'
+     *
+     * @throws InvalidArgumentException if an invalid attempt type is specified
+     * @return boolean
+     */
+    private function isWhitelisted($ip, $attemptType)
```
Comment (olleharstedt) [inline application/models/FailedLoginAttempt.php:150]: NB: This method does not contain `this`, so maybe it doesn't belong in a class...? But currently, there's no good system to put functions together with a domain entity, or "module".

File: application/helpers/sanitize_helper.php (hunk 2 above)
Comment (olleharstedt) [reply application/helpers/sanitize_helper.php:532]: To be a little anal, can you maybe do this as well?

> The first line in the docblock should be one sentence describing the relation between function arguments and its output

https://manual.limesurvey.org/Code_quality_guide#Contract

Comment (olleharstedt) [reply application/helpers/sanitize_helper.php:532]: > Checks the validity of an IPv4 address 

would be

> Checks the validity of an IPv4 address $input

Or maybe, with renaming the argument:

> Checks the validity of IPV4 address $ip

Comment (olleharstedt) [reply application/helpers/sanitize_helper.php:532]: Boolean arguments that changes the "mode" of the function can be hard to fit in a one-liner like that. :)

File: application/models/FailedLoginAttempt.php (hunk 3 above)
Comment (Shnoulle) [reply application/models/FailedLoginAttempt.php:150]: static function ?

Comment (olleharstedt) [reply application/models/FailedLoginAttempt.php:150]: Static functions cannot be mocked. But, sometimes they don't need to be mocked. :) (When they have no side-effects.)

This functions has one side-effect: It reads from app config. This should be lifted out if you want to unit-test this function.

Comment (Shnoulle) [reply application/models/FailedLoginAttempt.php:150]: read App->getConfig are side effect ???

Easy test : 
```
App()->setConfig("myconfigtotest","testvalue");
/* My test #1 */
App()->setConfig("myconfigtotest","testvalue2");
/* My test #2 */
```

Comment (olleharstedt) [reply application/models/FailedLoginAttempt.php:150]: Accessing globals are considered side-effects.

You can do setConfig, but then you have setup/teardown in the tests. Easy to forget, and tests can affect each other.

Comment (olleharstedt) [reply application/models/FailedLoginAttempt.php:150]: Oh, and it might be hard to run tests in parallel when you rely on global scope like that.

**Review by c-schmitz — CHANGES_REQUESTED** [review]
Please also consider the changes I requested in the 3.x PR for this issue

File: application/helpers/sanitize_helper.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/sanitize_helper.php:532]: > Boolean arguments that changes the "mode" of the function can be hard to fit in a one-liner like that

What do you mean? Sorry! Description for that argument would be too long?

**Review by c-schmitz — APPROVED** [review]

**Pull request merged**